src/askui/chat/api/messages/service.py

```diff
@@ -60,23 +81,229 @@ def create(
+                .filter(
+                    MessageOrm.id == leaf_id,
+                    MessageOrm.thread_id == thread_id,
+                    MessageOrm.workspace_id == workspace_id,
+                )
+                .cte(name="ancestors", recursive=True)
+            )
+
+            # Recursively traverse up until we hit ROOT_MESSAGE_PARENT_ID
+            _ancestors_recursive = select(MessageOrm.id, MessageOrm.parent_id).filter(
+                MessageOrm.id == _ancestors_cte.c.parent_id,
+                _ancestors_cte.c.parent_id != ROOT_MESSAGE_PARENT_ID,
+            )
+            _ancestors_cte = _ancestors_cte.union_all(_ancestors_recursive)
+
+            # Get the root node (the one with parent_id == ROOT_MESSAGE_PARENT_ID)
+            branch_root_id = self._session.execute(
+                select(MessageOrm.id).filter(
+                    MessageOrm.id.in_(select(_ancestors_cte.c.id)),
+                    MessageOrm.parent_id == ROOT_MESSAGE_PARENT_ID,
+                )
+            ).scalar_one_or_none()
+
+            if branch_root_id is None:
+                error_msg = f"Root message not found in thread {thread_id} for message {leaf_id}"
+                raise NotFoundError(error_msg)
+
+        else:
+            # Case 2: Set branch_root_id to query.before or ROOT node, then find latest leaf
+            if query.before:
+                branch_root_id = query.before
+            else:
+                # Get the latest root message
+                branch_root_id = self._session.execute(
+                    select(MessageOrm.id)
+                    .filter(
+                        MessageOrm.parent_id == ROOT_MESSAGE_PARENT_ID,
+                        MessageOrm.thread_id == thread_id,
+                        MessageOrm.workspace_id == workspace_id,
+                    )
+                    .order_by(desc(MessageOrm.id))
+                    .limit(1)
+                ).scalar_one_or_none()
+
+                # If no messages exist yet, return None
+                if branch_root_id is None:
+                    return None
+
+            # Build CTE to traverse down the tree from branch_root_id
+            _descendants_cte = (
+                select(MessageOrm.id, MessageOrm.parent_id)
+                .filter(
+                    MessageOrm.id == branch_root_id,
+                )
+                .cte(name="descendants", recursive=True)
+            )
+
+            # Recursively traverse down
+            _descendants_recursive = select(MessageOrm.id, MessageOrm.parent_id).filter(
+                MessageOrm.parent_id == _descendants_cte.c.id,
+            )
+            _descendants_cte = _descendants_cte.union_all(_descendants_recursive)
+
+            # Get the latest leaf (highest ID)
+            leaf_id = self._session.execute(
+                select(_descendants_cte.c.id)
+                .order_by(desc(_descendants_cte.c.id))
+                .limit(1)
+            ).scalar_one_or_none()
```

Comment:
Refactor:

Extract to method `_retrieve_latest_leaf(message_id: MessageId)` that uses `_build_decendents_cte()`